Basic fp8 support in Inductor

ipiszy opened the pull request:
Add basic fp8 support in Inductor, including:
* Fix fp8 Triton codegen issues;
* Add min_elements_per_thread requirement for fp8 related dtype conversions. More details on Triton implementation can be found from https://github.com/openai/triton/blob/10f59d8ce04052521c1bc0cb3a3f8b98918fc7e3/lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp#L10.

Note that the current implementation only works for Pointwise. Will create follow-up PRs for Reduction.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109765
* #109301
* __->__ #109168



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

ipiszy committed:
Basic fp8 support in Inductor

[ghstack-poisoned]

ipiszy committed:
Update on "Basic fp8 support in Inductor"

Add basic fp8 support in Inductor, including:
* Fix fp8 Triton codegen issues;
* Add min_elements_per_thread requirement for fp8 related dtype conversions. More details on Triton implementation can be found from https://github.com/openai/triton/blob/10f59d8ce04052521c1bc0cb3a3f8b98918fc7e3/lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp#L10.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "Basic fp8 support in Inductor"

Add basic fp8 support in Inductor, including:
* Fix fp8 Triton codegen issues;
* Add min_elements_per_thread requirement for fp8 related dtype conversions. More details on Triton implementation can be found from https://github.com/openai/triton/blob/10f59d8ce04052521c1bc0cb3a3f8b98918fc7e3/lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp#L10.

Note that the current implementation only works for Pointwise. Will create follow-up PRs for Reduction.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

drisspg commented on torch/utils/_sympy/value_ranges.py:
```diff
@@ -105,9 +112,13 @@ def __or__(self, other) -> "ValueRanges":
         if self.is_bool:
-            range = ValueRanges(sympy.And(self.lower, other.lower), sympy.Or(self.upper, other.upper))
+            range = ValueRanges(
+                sympy.And(self.lower, other.lower), sympy.Or(self.upper, other.upper)
```
These are all formatting correct?

ipiszy replied:
Fixed this PR to exclude formatting changes.

eellison reviewed (commented):
Would you mind separating the formatting prs ? it makes it more difficult to review. If you need configuring vscode or something else let me know.

Chillee commented on torch/_inductor/codegen/triton.py:
```diff
@@ -151,7 +155,42 @@ class TritonOverrides(OpOverrides):
+        def _get_min_elements_per_thread(
+            src_dtype: torch.dtype, dst_dtype: torch.dtype
+        ) -> int:
+            # fp8 data type conversions has min_elements_per_thread requirements.
```
Is there an explanation for why this is the case? 🤔

ipiszy replied:
Check the intrinsics in this file (also documented in the comment below): https://github.com/openai/triton/blob/10f59d8ce04052521c1bc0cb3a3f8b98918fc7e3/lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp#L10. It uses b32 and e4m3x2.

Chillee replied:
Or err, I saw the link, I'm just trying to get a better understanding for why dtype conversions would have min_elements_per_thread requirements.

ipiszy replied:
I think it's because ptx cvt only provides fp8*2 intrinsics, which are used by Triton: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#data-movement-and-conversion-instructions-cvt.
However, I also don't understand why the Triton implementation for fp8_e5m2 requires 4 element per thread. Asked in the Slack channel but didn't get an answer ^^.

Chillee commented on torch/_inductor/codegen/triton.py:
```diff
@@ -151,7 +155,42 @@ class TritonOverrides(OpOverrides):
 
     @staticmethod
-    def to_dtype(x, dtype: torch.dtype):
+    def to_dtype(x, dtype: torch.dtype, src_dtype: torch.dtype = None):
```
Can we not get `src_dtype` at this stage? Why do we need to plumb src_dtype through the entire lowering?

ipiszy replied:
This is because only fp8 conversion has special requirement on min_elements_per_thread. Any suggestions on better ways to implement this logic?

Chillee replied:
Yeah I was just a bit surprised that we don't have this information elsewhere during this lowering. I was thinking whether we can just get the dtype directly from `x`, but I think in general, we don't actually keep around dtype information at this stage of the lowering.

ipiszy commented:
> Would you mind separating the formatting prs ? it makes it more difficult to review. If you need configuring vscode or something else let me know.

Yeah will do. I was testing on another H100 host so maybe I did something wrong in the linter configuration...

ipiszy committed:
Update on "Basic fp8 support in Inductor"

Add basic fp8 support in Inductor, including:
* Fix fp8 Triton codegen issues;
* Add min_elements_per_thread requirement for fp8 related dtype conversions. More details on Triton implementation can be found from https://github.com/openai/triton/blob/10f59d8ce04052521c1bc0cb3a3f8b98918fc7e3/lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp#L10.

Note that the current implementation only works for Pointwise. Will create follow-up PRs for Reduction.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "Basic fp8 support in Inductor"

Add basic fp8 support in Inductor, including:
* Fix fp8 Triton codegen issues;
* Add min_elements_per_thread requirement for fp8 related dtype conversions. More details on Triton implementation can be found from https://github.com/openai/triton/blob/10f59d8ce04052521c1bc0cb3a3f8b98918fc7e3/lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp#L10.

Note that the current implementation only works for Pointwise. Will create follow-up PRs for Reduction.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "Basic fp8 support in Inductor"

Add basic fp8 support in Inductor, including:
* Fix fp8 Triton codegen issues;
* Add min_elements_per_thread requirement for fp8 related dtype conversions. More details on Triton implementation can be found from https://github.com/openai/triton/blob/10f59d8ce04052521c1bc0cb3a3f8b98918fc7e3/lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp#L10.

Note that the current implementation only works for Pointwise. Will create follow-up PRs for Reduction.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

drisspg commented on torch/_inductor/codegen/triton_utils.py:
```diff
@@ -10,7 +12,14 @@ def signature_of(arg: Union[TensorArg, SizeArg], *, size_dtype: str) -> str:
     if isinstance(arg, TensorArg):
-        tye = JITFunction._type_of(arg.dtype)
+        # TODO: Remove fp8 special handling when Triton supports PyTorch fp8 dtypes.
+        # Related PR: https://github.com/openai/triton/pull/2279/
```
Looks like this was merged but I imagine the triton version in pytorch hasn't been updated to have this commit

ipiszy replied:
Yeah I think we need to wait for Triton version update in Pytorch to remove these lines.

ipiszy committed:
Update on "Basic fp8 support in Inductor"

Add basic fp8 support in Inductor, including:
* Fix fp8 Triton codegen issues;
* Add min_elements_per_thread requirement for fp8 related dtype conversions. More details on Triton implementation can be found from https://github.com/openai/triton/blob/10f59d8ce04052521c1bc0cb3a3f8b98918fc7e3/lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp#L10.

Note that the current implementation only works for Pointwise. Will create follow-up PRs for Reduction.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "Basic fp8 support in Inductor"

Add basic fp8 support in Inductor, including:
* Fix fp8 Triton codegen issues;
* Add min_elements_per_thread requirement for fp8 related dtype conversions. More details on Triton implementation can be found from https://github.com/openai/triton/blob/10f59d8ce04052521c1bc0cb3a3f8b98918fc7e3/lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp#L10.

Note that the current implementation only works for Pointwise. Will create follow-up PRs for Reduction.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

drisspg reviewed (approved):
I think this looks good modulo linting but I think horrace + elias are a better review on this

eellison commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -843,20 +880,39 @@ def pointwise(size_hints, meta, tile_hint=None, filename=None):
-                triton_config(size_hints, bs, 1),
-                triton_config(size_hints, 1, bs),
+                triton_config(
+                    size_hints, 32, 32, min_elements_per_thread=min_elements_per_thread
```
nit: doesn't really matter, but maybe slightly less verbose as `min_elem_per_thread`

eellison commented on torch/_inductor/triton_heuristics.py:
```diff
@@ -820,7 +843,11 @@ def pointwise(size_hints, meta, tile_hint=None, filename=None):
         ):
```
for fewer changes you could consider `triton_config = functools.partial(triton_config, min_elements_per_thread=min_elements_per_thread`

eellison commented on torch/_prims_common/__init__.py:
```diff
@@ -830,7 +830,14 @@ def infer_size(shape: ShapeType, numel: int) -> Tuple[int, ...]:
 _integer_dtypes = (torch.uint8, torch.int8, torch.int16, torch.int32, torch.int64)
 _low_precision_dtypes = (torch.float16, torch.bfloat16, torch.complex32)
-_float_dtypes = (torch.float16, torch.bfloat16, torch.float32, torch.float64)
+_float_dtypes = (
```
Can we replace this with calling `dtype.is_floating_point`

eellison reviewed (commented):
Do we need to update reduction heuristics as well ? What about triton templates ?

aakhundov commented on test/inductor/test_fp8.py:
```diff
@@ -56,10 +44,54 @@ def fp8_matmul_unwrapped(x):
+
+        x_shape = (16, 16, 16)
+
+        with self.assertRaises(Exception):
```
Would it make sense to be more specific here re. the `Exception` type (and maybe text)? And below.

ipiszy committed:
Update on "Basic fp8 support in Inductor"

Add basic fp8 support in Inductor, including:
* Fix fp8 Triton codegen issues;
* Add min_elements_per_thread requirement for fp8 related dtype conversions. More details on Triton implementation can be found from https://github.com/openai/triton/blob/10f59d8ce04052521c1bc0cb3a3f8b98918fc7e3/lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp#L10.

Note that the current implementation only works for Pointwise. Will create follow-up PRs for Reduction.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy reviewed (commented):
Thanks @eellison @aakhundov !

@eellison Changes for reduction heuristics will be in the next PR. I'm still trying to figure out how to do it correctly for different reduction types (persistent_reduction and normal reduction, real reduction (like max) and fused_reduction_pointwise (like layer_norm)).

Triton template changes will come the last. For now we'll just rely on CuBLAS for Gemms since Triton H100 perf is not ideal.

ipiszy committed:
Update on "Basic fp8 support in Inductor"

Add basic fp8 support in Inductor, including:
* Fix fp8 Triton codegen issues;
* Add min_elements_per_thread requirement for fp8 related dtype conversions. More details on Triton implementation can be found from https://github.com/openai/triton/blob/10f59d8ce04052521c1bc0cb3a3f8b98918fc7e3/lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp#L10.

Note that the current implementation only works for Pointwise. Will create follow-up PRs for Reduction.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy committed:
Update on "Basic fp8 support in Inductor"

Add basic fp8 support in Inductor, including:
* Fix fp8 Triton codegen issues;
* Add min_elements_per_thread requirement for fp8 related dtype conversions. More details on Triton implementation can be found from https://github.com/openai/triton/blob/10f59d8ce04052521c1bc0cb3a3f8b98918fc7e3/lib/Conversion/TritonGPUToLLVM/ElementwiseOpToLLVM.cpp#L10.

Note that the current implementation only works for Pointwise. Will create follow-up PRs for Reduction.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ipiszy commented:
I'll merge this PR first to unblock fp8 related testing. Meanwhile, I'm working on adding scalar fp8 conversion support in the Triton repo, and re-visit the Pointwise TritonHeuristics related change after the fix.

ipiszy commented:
@pytorchbot merge

ipiszy commented:
@pytorchbot label "topic: not user facing"

ipiszy commented:
@pytorchbot merge
